contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -5972,3 +5978,47 @@ allow_storing_init_privs(Oid objoid, Oid classoid, int objsubid)
 	}
 	return true;
 }
+
+/*
+ * remove_db_name_in_schema - remove the db name and underscore at the beginning
+ * 	of the given string. It is used to unmap schema name in error messages.
+ *
+ * 	@param schema_name - char *
+ * 	@return - unmapped schema name char *
+ */
+static const char *
+remove_db_name_in_schema(const char *object_name)
+{
+	char		*cur_db_name;
+	char		**splited_object_name;
+	char		*schema_name = NULL;
+	char		*mutable_name;
+    size_t		db_name_len;
+    size_t		prefix_len = 0;
+	size_t		schema_name_len;
+
+	mutable_name = pstrdup(object_name);
+	splited_object_name = split_object_name(mutable_name);
+
+	/* If there are more than two parts, then it’s a cross-db object name (db.schema.object) so we don’t need to deal with it */
+	if (strlen(splited_object_name[1]) == 0)
+	{
+		schema_name = strlen(splited_object_name[2]) == 0 ? splited_object_name[3] : splited_object_name[2];
+		cur_db_name = get_cur_db_name();
+		db_name_len = strlen(cur_db_name);
+		schema_name_len = strlen(schema_name);
+
+		if (schema_name != NULL && schema_name_len > db_name_len && strncmp(schema_name, cur_db_name, db_name_len) == 0 && schema_name[db_name_len] == '_') {
```

Comment:
We can look at the babelfish catalog for original schema name instead. we have syscache for it as well. `SYSNAMESPACENAME`